aten/src/ATen/native/cpu/ReduceOpsKernel.cpp

```diff
@@ -174,61 +174,72 @@ static void norm_kernel_tensor_iterator_impl(
   if (p.isIntegral(false)) {
     val = p.to<int64_t>();
   } else if (p.isFloatingPoint()) {
-    val = p.to<float>();
+    val = p.to<double>();
   } else {
     AT_ERROR("norm_kernel_tensor_iterator_impl expects norm to be integer or float");
   }
 
-
   if (val == 0) {
     AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES_AND2(kHalf, kBFloat16, iter.dtype(), "norm_cpu", [&] {
+      using acc_t = typename scalar_value_type<scalar_t>::type;
       binary_kernel_reduce(
         iter,
-        NormZeroOps<scalar_t>(),
-        scalar_t(0)
+        NormZeroOps<scalar_t, acc_t>(),
+        acc_t(0)
       );
     });
   } else if (val == 1) {
     AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES_AND2(kHalf, kBFloat16, iter.dtype(), "norm_cpu", [&] {
+      using acc_t = typename scalar_value_type<scalar_t>::type;
       binary_kernel_reduce(
         iter,
-        NormOneOps<scalar_t>(),
-        scalar_t(0)
+        NormOneOps<scalar_t, acc_t>(),
+        acc_t(0)
       );
     });
   } else if (val == 2) {
     AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES_AND2(kHalf, kBFloat16, iter.dtype(), "norm_cpu", [&] {
+      using acc_t = typename scalar_value_type<scalar_t>::type;
       binary_kernel_reduce(
         iter,
-        NormTwoOps<scalar_t>(),
-        scalar_t(0)
+        NormTwoOps<scalar_t, acc_t>(),
+        acc_t(0)
       );
     });
   } else if (val == INFINITY) {
     AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES_AND2(kHalf, kBFloat16, iter.dtype(), "norm_cpu", [&] {
+      using acc_t = typename scalar_value_type<scalar_t>::type;
       binary_kernel_reduce(
         iter,
-        AbsMaxOps<scalar_t>(),
-        scalar_t(std::numeric_limits<scalar_t>::min())
+        AbsMaxOps<scalar_t, acc_t>(),
+        std::numeric_limits<acc_t>::min()
       );
     });
   } else if (val == -INFINITY) {
     AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES_AND2(kHalf, kBFloat16, iter.dtype(), "norm_cpu", [&] {
+      using acc_t = typename scalar_value_type<scalar_t>::type;
       binary_kernel_reduce(
         iter,
-        AbsMinOps<scalar_t>(),
-        scalar_t(std::numeric_limits<scalar_t>::max())
+        AbsMinOps<scalar_t, acc_t>(),
+        std::numeric_limits<acc_t>::max()
       );
     });
   } else {
     AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES_AND2(kHalf, kBFloat16, iter.dtype(), "norm_cpu", [&] {
+      using acc_t = typename scalar_value_type<scalar_t>::type;
       binary_kernel_reduce(
         iter,
-        NormOps<scalar_t> { scalar_t(val) },
-        scalar_t(0)
+        NormOps<scalar_t, acc_t> { acc_t(val) },
+        acc_t(0)
       );
     });
   }
+
+  // For complex outputs, the above kernels do not touch the imaginary values,
```

Comment:
Would you elaborate on this? How do the above kernels not edit the imaginary values of output?

I wonder if it's faster to zero output before it's modified instead of zeroing just the imaginary part after its been modified.

Reply:
Sure, I can elaborate. I'll try zeroing the whole thing first and see how it affects performance.

Reply:
Actually this special case won't be necessary any more after I change the output types to match numpy.

Reply:
Nevermind, I misspoke. This is special case is still important if `out`'s dtype or the `dtype` argument is specified as complex.

Reply:
Ok, I looked into whether zeroing out the entire tensor before the norm calculation would be faster than zeroing out only the imaginary part after the norm calculation. I used this script: https://github.com/kurtamohler/pytorch-perf-test-scripts/blob/master/linalg_norm/run_single_core_norm_output_complex_perf.sh

There doesn't seem to be a significant difference between the two options.

Reply:
Cool. Thanks for checking, @kurtamohler!